### PR TITLE
Fix part of #2502: Fix text input layout tests

### PR DESCRIPTION
**Pull request by FareesHussain**
## Explanation

Fixes part of #2502

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### Espresso Tests for `AddProfileActivityTest`:
image

### Espresso Tests for `ProfileRenameActivityTest`:
image

### Espresso Tests for `ProfileResetPinActivityTest`:
image

### Espresso Tests for `AdminAuthActivityTest`:
image

### CODE REVIEW

File: app/src/main/java/org/oppia/android/app/profile/AddProfileActivity.kt [hunk 1]
```diff
@@ -44,4 +44,14 @@ class AddProfileActivity : InjectableAppCompatActivity() {
...
+
+  override fun onSaveInstanceState(outState: Bundle) {
+    addProfileFragmentPresenter.handleOnSavedInstanceState(outState)
+    super.onSaveInstanceState(outState)
+  }
+
+  override fun onRestoreInstanceState(savedInstanceState: Bundle?) {
+    super.onRestoreInstanceState(savedInstanceState)
+    addProfileFragmentPresenter.handleOnRestoreInstanceState(savedInstanceState!!)
+  }
```
Comment (prayutsu) [inline app/src/main/java/org/oppia/android/app/profile/AddProfileActivity.kt]: Using `onSaveInstanceState` we can survive configuration changes, but what if the app keeps running for a long time in the background(maybe for a day), and then we open the app from "recent apps", will this still work?

Comment (FareesHussain) [reply app/src/main/java/org/oppia/android/app/profile/AddProfileActivity.kt]: @prayutsu I don't understand what your doubt is
In most cases when the app is in the background for a day the app will restart
else there will be no issues but it has nothing to do with onSavedInstance

Correct me if I'm wrong

Comment (prayutsu) [reply app/src/main/java/org/oppia/android/app/profile/AddProfileActivity.kt]: Actually, I was more specifically talking about `System-initiated process death` and I think `onSavedInstanceState()` survives it easily.

**Comment by prayutsu** [comment]
@FareesHussain You can mark the above conversation as resolved.

**Review by anandwana001 — CHANGES_REQUESTED** [review]
@FareesHussain Why the test was passing without this implementation earlier?

**Comment by FareesHussain** [comment]
> @FareesHussain Why the test was passing without this implementation earlier?
> 

@anandwana001 it was working before we implemented the databinding for the input and error message, I've run the tests in that branch and these tests fail there too.

**Comment by anandwana001** [comment]
Could you please add that commit message?

**Comment by FareesHussain** [comment]
https://github.com/oppia/oppia-android/commit/16f51a5cd1a0bb82a94703b1c835b2f06fc7fc7b

**Comment by anandwana001** [comment]
We need to fix this with Databinding. Need to look for a solution for this. @FareesHussain

**Comment by FareesHussain** [comment]
> We need to fix this with Databinding. Need to look for a solution for this. @FareesHussain

I've tried to use the following code similar to `ResetPinDialog`

```kotlin

    binding.addProfileActivityUserNameEditText.onTextChanged { name ->
      name?.let {
        if(
          profileViewModel.nameErrorMsg.get()?.isNotEmpty()!! &&
          profileViewModel.inputName.get() == it
        ) {
          profileViewModel.inputName.set(it)
          profileViewModel.isButtonActive.set(it.isNotEmpty())
        } else {
          profileViewModel.isButtonActive.set(it.isNotEmpty())
          profileViewModel.nameErrorMsg.set("")
          profileViewModel.inputName.set(it)
        }
      }
    }

    binding.addProfileActivityPinEditText.onTextChanged { pin ->
      pin?.let {
      if(profileViewModel.pinErrorMsg.get()?.isNotEmpty()!! &&
        profileViewModel.inputPin.get() == it
      ) {
        profileViewModel.inputPin.set(it)
      } else {
          profileViewModel.inputPin.set(it)
          profileViewModel.pinErrorMsg.set("")
          inputtedPin = pin.isNotEmpty()
          setValidPin(binding)
        }
      }
    }

    binding.addProfileActivityConfirmPinEditText.onTextChanged { confirmPin ->
      confirmPin?.let {
        if (
          profileViewModel.confirmPinErrorMsg.get()?.isNotEmpty()!! &&
          profileViewModel.inputConfirmPin.get() == it
        ) {
          profileViewModel.inputConfirmPin.set(it)
        } else {
          profileViewModel.inputConfirmPin.set(it)
          profileViewModel.confirmPinErrorMsg.set("")
          inputtedConfirmPin = confirmPin.isNotEmpty()
          setValidPin(binding)
        }
      }
    }

```

**Comment by anandwana001** [comment]
does it work? 

Though, we need to look for solutions if we can do this with data binding.

**Comment by FareesHussain** [comment]
> does it work?
> 
> Though, we need to look for solutions if we can do this with data binding.

@anandwana001 let me be a bit clear 

The main issue here is the error message in `TextInputLayout` disappears after the orientation change

This issue was not present when we created the Pull request and all tests were passing but after we included the data binding this issue appeared and the Espresso was not run

and to solve this we can either use `onSavedInstance` or the code I've mentioned in the above comment https://github.com/oppia/oppia-android/pull/2510#issuecomment-762653179 and we are already using databinding here

What is your opinion on this?

**Comment by anandwana001** [comment]
The code mentioned above sounds good to me. Let's update the PR and send for review.

**Comment by FareesHussain** [comment]
> The code mentioned above sounds good to me. Let's update the PR and send for review.

@anandwana001 As we need to maintain consistency for all Activities related to TextInputLayout shall I update `AdminAuthActivity` removing `onSavedInstance`  in this PR

**Comment by anandwana001** [comment]
Are those related to this comment - https://github.com/oppia/oppia-android/issues/2502#issuecomment-762767756 ?

**Comment by FareesHussain** [comment]
> Are those related to this comment - [#2502 (comment)](https://github.com/oppia/oppia-android/issues/2502#issuecomment-762767756) ?

What is mean is we are using `onSavedInstance` for `AdminAuthActivity` as we are using the above code to solve this issue shall I update `AdminAuthActivity` as well?

**Comment by anandwana001** [comment]
yeah, sure.

**Comment by prayutsu** [comment]
If we decide to finalize this approach, then I can also fix part of #2510 (AdminPinActivityTests) 

> 
> I've tried to use the following code similar to `ResetPinDialog`
> 
> ```kotlin
>     binding.addProfileActivityUserNameEditText.onTextChanged { name ->
>       name?.let {
>         if(
>           profileViewModel.nameErrorMsg.get()?.isNotEmpty()!! &&
>           profileViewModel.inputName.get() == it
>         ) {
>           profileViewModel.inputName.set(it)
>           profileViewModel.isButtonActive.set(it.isNotEmpty())
>         } else {
>           profileViewModel.isButtonActive.set(it.isNotEmpty())
>           profileViewModel.nameErrorMsg.set("")
>           profileViewModel.inputName.set(it)
>         }
>       }
>     }
> 
>     binding.addProfileActivityPinEditText.onTextChanged { pin ->
>       pin?.let {
>       if(profileViewModel.pinErrorMsg.get()?.isNotEmpty()!! &&
>         profileViewModel.inputPin.get() == it
>       ) {
>         profileViewModel.inputPin.set(it)
>       } else {
>           profileViewModel.inputPin.set(it)
>           profileViewModel.pinErrorMsg.set("")
>           inputtedPin = pin.isNotEmpty()
>           setValidPin(binding)
>         }
>       }
>     }
> 
>     binding.addProfileActivityConfirmPinEditText.onTextChanged { confirmPin ->
>       confirmPin?.let {
>         if (
>           profileViewModel.confirmPinErrorMsg.get()?.isNotEmpty()!! &&
>           profileViewModel.inputConfirmPin.get() == it
>         ) {
>           profileViewModel.inputConfirmPin.set(it)
>         } else {
>           profileViewModel.inputConfirmPin.set(it)
>           profileViewModel.confirmPinErrorMsg.set("")
>           inputtedConfirmPin = confirmPin.isNotEmpty()
>           setValidPin(binding)
>         }
>       }
>     }
> ```

**Comment by FareesHussain** [comment]
I've updated the description with Espresso test screenshots for the recent changes

File: app/src/main/java/org/oppia/android/app/profile/AdminAuthViewModel.kt [hunk 2]
```diff
@@ -10,6 +10,7 @@ import javax.inject.Inject
 /** The ViewModel for [AdminAuthActivity]. */
 @ActivityScope
 class AdminAuthViewModel @Inject constructor() : ObservableViewModel() {
+  val inputPin = ObservableField("")
```
Comment (anandwana001) [inline app/src/main/java/org/oppia/android/app/profile/AdminAuthViewModel.kt]: `ObservableField<String>`

**Comment by FareesHussain** [comment]
@anandwana001 PTAL

**Review by anandwana001 — APPROVED** [review]
nit change. 
LGTM

File: app/src/main/java/org/oppia/android/app/profile/AdminAuthViewModel.kt (hunk 2 above)
Comment (FareesHussain) [reply app/src/main/java/org/oppia/android/app/profile/AdminAuthViewModel.kt]: Done

**Comment by FareesHussain** [comment]
@rt4914 PTAL

**Review by rt4914 — APPROVED** [review]
LGTM, thanks.

**Pull request merged**